### PR TITLE
Update Apache Lucene to 9.8.0-snapshot-4373c3b

### PR DESCRIPTION
**Pull request by reta**
### Description
Update Apache Lucene to 9.8.0-snapshot-79b9664

### Related Issues
N/A

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
Test failures are unrelated: https://github.com/opensearch-project/OpenSearch/issues/8699

File: server/src/main/java/org/opensearch/search/suggest/completion/CompletionSuggester.java [hunk 1]
```diff
@@ -116,6 +116,7 @@ private static void suggest(IndexSearcher searcher, CompletionQuery query, TopSu
                 }
             }
         }
+        collector.finish();
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/suggest/completion/CompletionSuggester.java:119]: See please https://github.com/apache/lucene/pull/12380

**Review by kotwanikunal — APPROVED** [review]

**Comment by reta** [comment]
@nknize mind taking a quick look as well please?

**Comment by reta** [comment]
> @reta @nknize Will this lucene version change in 2.11 release? This is corresponding to an important decision-making in our [feature](https://github.com/opensearch-project/neural-search/issues/284). This is a re-invent feature and scheduled to release at 2.11.

@zhichao-aws if 9.8.0 happens to be released before 2.11, I think we could get it in, but I don't have a date for this release sadly [1]

[1] https://github.com/apache/lucene/milestones

**Comment by zhichao-aws** [comment]
> > @reta @nknize Will this lucene version change in 2.11 release? This is corresponding to an important decision-making in our [feature](https://github.com/opensearch-project/neural-search/issues/284). This is a re-invent feature and scheduled to release at 2.11.
> 
> @zhichao-aws if 9.8.0 happens to be released before 2.11, I think we could get it in, but I don't have a date for this release sadly [1]
> 
> [1] https://github.com/apache/lucene/milestones

I see, thanks!

**Comment by msfroh** [comment]
For the record, @zhaih kicked off the 9.8 release discussion on the Lucene mailing list a couple of days ago: https://lists.apache.org/thread/xvhqdb0vzyfqsbpl9lvlrxxm2v4tr39z

**Comment by zhichao-aws** [comment]
> @zhichao-aws For the record, @zhaih kicked off the 9.8 release discussion on the Lucene mailing list a couple of days ago: https://lists.apache.org/thread/xvhqdb0vzyfqsbpl9lvlrxxm2v4tr39z

That's good to know, thanks :)

**Comment by zhichao-aws** [comment]
Lucene has cut the 9.8 branch today. Can we make it to upgrade to lucene 9.8 in 2.x before 2.11 release?

**Comment by reta** [comment]
> Lucene has cut the 9.8 branch today https://github.com/apache/lucene/tree/branch_9_8. Can we make it to upgrade to lucene 9.8 in 2.x before 2.11 release?

Once the artifacts are published to Maven Central - for sure

**Comment by CEHENKLE** [comment]
Hey @reta We've got this weird little "hurry up" release locking next week (Oct 4th), and I'm worried we won't have a chance to absorb these changes with everything else going on.  What would you think about holding merging this until after Oct 18th?  Is there a reason we should to it sooner?

**Comment by reta** [comment]
> What would you think about holding merging this until after Oct 18th? Is there a reason we should to it sooner?

Hey @CEHENKLE , this was not backported to `2.x` (we backport only releases), so non-issue  :)

**Comment by CEHENKLE** [comment]
Ah coolio.  Brainfail on my side.  Thank you, @reta!

**Comment by CEHENKLE** [comment]
Hey, it looks like neural-search may be need these changes for 2.11 (I've asked them to comment on the issue).  In the meantime, @reta (or others), do you know how impactful these changes would be to other parts of the core/the plugins if we merged them?  

It's hard, because if we're going to do it, we need to go as soon as possible to give time for folks to absorb changes.  But if we're not going to go, I don't want to backport and cause thrash. :( 

(@nknize, @andrross for more eyes)

**Comment by msfroh** [comment]
Technically this PR won't be the one to be backported, but rather https://github.com/opensearch-project/OpenSearch/pull/10276 (based on the released Lucene 9.8.0, which may be identical to the snapshot merged in this PR).

The differences likely to impact other parts of the code are probably the API changes: https://lucene.apache.org/core/9_8_0/changes/Changes.html#v9.8.0.api_changes, and changes to runtime behavior: https://lucene.apache.org/core/9_8_0/changes/Changes.html#v9.8.0.changes_in_runtime_behavior

Those changes to runtime behavior would potentially seem risky for our concurrent search work, but AFAIK, we're not seeing differences between `2.x` and `main` -- right? (If anything, ensuring that 2.x and main behave the same there feels like that will reduce thrashing.)

**Comment by reta** [comment]
> Technically this PR won't be the one to be backported, but rather #10276 (based on the released Lucene 9.8.0, which may be identical to the snapshot merged in this PR).

@CEHENKLE as @msfroh pointed out, Apache Lucene 9.8.0 is released so technically we are good to go for 2.x (probably make it to 2.11). We don't see any regressions so far, the changes related to concurrent search seems to be properly tracked and accommodated.

**Comment by model-collapse** [comment]
Thanks @msfroh and @reta for pointing out. From neural-search side, Lucene 9.8 is needed so as not to overwirite the FeatureField implementation for performance issues and we don't want to deprecate in the future the parameter (for performance tuning) which we have to bring up in the condition of Lucene 9.7. 

Meanwhile we are also aware of some potential risks of pushing [#10276](https://github.com/opensearch-project/OpenSearch/pull/10276) forward: 
1) Before OpenSearch 2.11, the main branch of ml-commons is behind 2.x on some commits. Although this has fixed at the moment, we still suspect that some other repos may have the same kind of problems. The main branch should catch up with 2.x with the latest features & implementations before we back-port Lucene 9.8 to  2.x, otherwise, there will be endless conflicts to deal with.
2) We believe that all repos have done their functional tests as well as some benchmarkings to their main branch with Lucene 9.8 snapshot and so far there was no negative feedback. But as a rigorous procedure, we suggest all the repo to reproduce their tests with the formal release artifacts of Luence 9.8 (not snapshot version) so as to validate whether there's some negative impact from the Lucene upgrade to OpenSearch. The neural search plugin is absorbing the performance improvement, we also want to see others' feedback.
3) We'd like to see whether the change of [APIs](https://lucene.apache.org/core/9_8_0/changes/Changes.html#v9.8.0.api_changes) or [runtime behavior](https://lucene.apache.org/core/9_8_0/changes/Changes.html#v9.8.0.changes_in_runtime_behavior) will have impact to AWS hosted search service which will subscribe this upgrade in the feature. That is another topic. We shall be aware but not discuss it on github.

**Comment by reta** [comment]
> Meanwhile we are also aware of some potential risks of pushing [#10276](https://github.com/opensearch-project/OpenSearch/pull/10276) forward:

Thanks @model-collapse , with concerns you have raised it looks like we should not rush 9.8.0 into `2.11` branch (since the release is a few days away) but wait for cut off and plan for `2.12`.

**Comment by Pallavi-AWS** [comment]
@model-collapse do you agree with below assessment from @reta? With the tight timelines for 2.11 release, ideally we don't want to force a Lucene upgrade. 

> > Meanwhile we are also aware of some potential risks of pushing [#10276](https://github.com/opensearch-project/OpenSearch/pull/10276) forward:
> 
> Thanks @model-collapse , with concerns you have raised it looks like we should not rush 9.8.0 into `2.11` branch (since the release is a few days away) but wait for cut off and plan for `2.12`.

**Comment by krishna-ggk** [comment]
> I agree with @msfroh that most repo will follow same pattern of commit to main and backport to 2.x in general. Therefore, I don't see a much risk on point 1 and point 2 especially we have been using lucene 9.8 snapshot in main branch for a while.

IMHO, I don't think we can equate testing of main to be equal to testing for 2.x. Just a [simple diff](https://github.com/opensearch-project/OpenSearch/compare/main...2.x) (without any contextual/functional analysis) shows 5k+ files, 1400+ commits and 70k+ lines. Agree with assessment, lucene upgrade nearing 2.11 release window does feel risky.

**Pull request merged**